### PR TITLE
`tf.keras` version that allows any input resolution and doesn't use `Lambda` layers

### PR DESCRIPTION
**Pull request by prhbrt**
Since the UNET architecture only uses layers that can scale with the image dimensions, the fixed dimensions seem artificial. I've added a zero-padding layer that increases the dimensions to the nearest multiple of 32. The padding is cut off in the end.

Moreover, I've removed the lambda-layer, as it creates marshaling warning, and used `ZeroPadding2D`'s asymmetric padding feature. This removes a warning upon `load_model`.

I [converted the eynollah-models to this architecture](https://colab.research.google.com/drive/1TA9euC_p0PAdYIsWOFJk6CcTBa9JZXeM#scrollTo=kZ0RpJ8Oly_-), and they should load without warnings now and use the `tensorflow.keras` API and can be found [here](https://drive.google.com/file/d/1JXJmxPYRCHnvBhKeoxbkFgjDJG9iDTsC/view?usp=sharing).

This might allow you to skip the patching as used in Eynollah and speed up the whole process. Please let me know what you think.

Notes and sanity checks:

 * the enhancement model uses the light version without the last batchnorm and with a sigmoid rather than softmax.
 * the column classifier is a different architecture, I just copied the original model, and hence it's not dimension invariant.
 * Two models weren't in savemodel-format on your website, e.g. the light versions `eynollah-main-regions_20220314` and `eynollah-textline_light_20210425`, so I couldn't convert them.

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -1,22 +1,12 @@
...
-from keras import layers
-from keras.regularizers import l2
+from tensorflow.keras.models import Model
+from tensorflow.keras.layers import Conv2D, Concatenate, ZeroPadding2D, BatchNormalization, Activation, MaxPooling2D, UpSampling2D, Input, Layer
+from tensorflow.keras import layers
+from tensorflow.keras.regularizers import l2
+import tensorflow as tf
 
 resnet50_Weights_path='./pretrained_model/resnet50_weights_tf_dim_ordering_tf_kernels_notop.h5'
-IMAGE_ORDERING ='channels_last'
```
Comment (prhbrt) [inline models.py:7]: This should be passed as a variable, and ideally use tensorflow's default.

TODO: default to `image_data_format` from keras' config (`~/.keras/keras.json`)

File: models.py [hunk 2]
```diff
@@ -1,22 +1,12 @@
...
-MERGE_AXIS=-1
```
Comment (prhbrt) [inline models.py:8]: This should follow from `data_format`, e.g. should always be the channel-axis.

File: models.py [hunk 3]
```diff
@@ -1,22 +1,12 @@
...
-def one_side_pad( x ):
-    x = ZeroPadding2D((1, 1), data_format=IMAGE_ORDERING)(x)
-    if IMAGE_ORDERING == 'channels_first':
-        x = Lambda(lambda x : x[: , : , :-1 , :-1 ] )(x)
-    elif IMAGE_ORDERING == 'channels_last':
-        x = Lambda(lambda x : x[: , :-1 , :-1 , :  ] )(x)
-    return x
-
-def identity_block(input_tensor, kernel_size, filters, stage, block):
+def identity_block(input_tensor, kernel_size, filters, stage, block, data_format='channels_last'):
```
Comment (prhbrt) [inline models.py:9]: Added `data_format` as a parameter everywhere, rather than the global constant `IMAGE_FORMAT`. Also used the same name as `tf` does for transparency.

File: models.py [hunk 4]
```diff
@@ -99,219 +83,131 @@ def conv_block(input_tensor, kernel_size, filters, stage, block, strides=(2, 2))
...
-    img_input = Input(shape=(input_height,input_width , 3 ))
-
-    if IMAGE_ORDERING == 'channels_last':
-        bn_axis = 3
-    else:
-        bn_axis = 1
-
-    x = ZeroPadding2D((3, 3), data_format=IMAGE_ORDERING)(img_input)
-    x = Conv2D(64, (7, 7), data_format=IMAGE_ORDERING, strides=(2, 2),kernel_regularizer=l2(weight_decay), name='conv1')(x)
+class PadMultiple(Layer):
```
Comment (prhbrt) [inline models.py:86]: This pads to a multiple of `32` or whatever is specified in `dims`.

File: models.py [hunk 5]
```diff
@@ -99,219 +83,131 @@ def conv_block(input_tensor, kernel_size, filters, stage, block, strides=(2, 2))
...
+        return inputs[0][:, :h, :w] if self.data_format == 'channels_last' else inputs[0][:, :, :h, :w]
+
+
+def resnet50_unet(n_classes, input_height=None, input_width=None, weight_decay=1e-6, pretraining=False, last_activation='softmax', skip_last_batchnorm=False, light_version=False, data_format='channels_last'):
+    """ Returns a U-NET model using the keras functional API. """
+    img_input = Input(shape=(input_height, input_width, 3 ))
+    padded_to_multiple = PadMultiple((32,32))(img_input)
+
+    bn_axis = 3 if data_format == 'channels_last' else 1
+    merge_axis = 3 if data_format == 'channels_last' else 1
```
Comment (prhbrt) [inline models.py:117]: Is `merge_axis` ever something else than the channel dimension?

File: models.py [hunk 6]
```diff
@@ -99,219 +83,131 @@ def conv_block(input_tensor, kernel_size, filters, stage, block, strides=(2, 2))
...
+    x = identity_block(x, 3, [512, 512, 2048], stage=5, block='b', data_format=data_format)
+    x = identity_block(x, 3, [512, 512, 2048], stage=5, block='c', data_format=data_format)
     f5 = x 
 
-
     if pretraining:
-        model=Model( img_input , x ).load_weights(resnet50_Weights_path)
+        Model(img_input, x).load_weights(resnet50_Weights_path)
 
+    if light_version:
```
Comment (prhbrt) [inline models.py:156]: Duplicate code complicates code maintenance, so I merged the light version in here.

**Comment by cneud** [comment]
Hi @prhbrt, thanks a lot for looking into this and for contributing! 

FYI, we are planning to update and refactor this repo and integrate the model training code with https://github.com/qurator-spk/eynollah for future maintenance, so this comes in very handy. 

My colleagues @vahidrezanezhad and @michalbubula will be working on this - although due to various reasons, we likely won't be able to get our hands dirty much before March. But we will try our best to review and merge any contributions also beforehand. 

Btw the two models that you were missing should be available from our [HF](https://huggingface.co/SBB):
* `eynollah-main-regions_20220314` -> https://huggingface.co/SBB/eynollah-main-regions
* `eynollah-textline_light_20210425` -> https://huggingface.co/SBB/eynollah-textline_light

**Comment by prhbrt** [comment]
@cneud Understood! Could you in the meantime provide a list of all model-architectures used for eynollah (specifically python code)? I couldn't find python-code for the column classifier in particular, so that one still has fixed dimensions.

Also note that this yolo-version might give slightly different outputs as your patching example, due to boundary conditions.

Thank you in advance!

**Comment by vahidrezanezhad** [comment]
> @cneud Understood! Could you in the meantime provide a list of all model-architectures used for eynollah (specifically python code)? I couldn't find python-code for the column classifier in particular, so that one still has fixed dimensions.
> 
> Also note that this yolo-version might give slightly different outputs as your patching example, due to boundary conditions.
> 
> Thank you in advance!

@prhbrt  
Sure, I'll try to make classifier code public in the meantime.

**Comment by kba** [comment]
I have tried to port this over to eynollah here: https://github.com/qurator-spk/eynollah/pull/202

If I messed it up, this PR is still around but I am closing so we can archive the repository.
